runtime/lua/vim/ui.lua

```diff
@@ -0,0 +1,32 @@
+local M = {}
+
+--- Prompts the user to pick a single item from a collection of entries
+---
+---@param entries table Arbitrary entries
+---@param opts table Additional options
+---     - prompt (string|nil)
+---               Text of the prompt. Defaults to `Pick one of:`
+---     - format_entry (function entry -> text)
+---               Function to format an
+---               individual entry from `entries`. Defaults to `tostring`.
+---@param callback function ((entry|nil, idx|nil) -> ())
+---               Called once the user made a choice.
+---               `idx` is the 1-based index of `entry` within `entries`.
+---               `nil` if the user aborted the dialog.
+function M.pick_one(entries, opts, callback)
+  opts = opts or {}
+  local choices = {opts.prompt or 'Pick one of:'}
+  local format_entry = opts.format_entry or tostring
+  for i, item in pairs(entries) do
+    table.insert(choices, string.format('%d: %s', i, format_entry(item)))
+  end
```

Comment:
If we wanted, we could support `entries` being a function following the iterator protocol. It would make 3rd party `pick_one` implementations slightly more complex, but it would enable support for providing the data lazily, avoiding the need to fully materialize a table with all items.